File: LeetCode/LeetCode-CN-JAVA/2020六月每日一题/1300. 转变数组后最接近目标值的数组和.java
class Solution {
    public int findBestValue(int[] arr, int target) {
        // C++ 用的二分，这里使用平均值
        Arrays.sort(arr); // 首先进行排序
        int sum = 0;
        for(int i = 0 ; i < arr.length; i++) {
            int x = (target - sum) / (arr.length - i); // 计算出期望值与实际值的平均值
            if(arr[i] > x) { // 由于是递增关系，如果当前的数值比平均值还大，则不可能更接近了
                double r =  (double)(target - sum) / (arr.length - i);
                return r - x > 0.5 ? x + 1 : x;
            }
            sum += arr[i];
        }
        return arr[arr.length - 1]; // 说明数组中所有的元素都小于等于评平均值，那么数组不需要改动，因为只会越改越小
    }
}

File: LeetCode/LeetCode-CN-CPP/2020六月每日一题/1300. 转变数组后最接近目标值的数组和.cpp
class Solution {
public:
    int findBestValue(vector<int>& arr, int target) {
        sort(arr.begin(), arr.end()); // 先进行排序
        vector<int> prefix; // prefix 的i保存的是arr中i及i之前的和
        for(int i = 0; i < arr.size(); i++) {
            int sub_sum = i == 0 ? arr[i] : arr[i] + prefix[i-1];
            prefix.push_back(sub_sum);
            //cout << prefix[i] << " ";
        }
        
        int ans = 0;
        int max = *max_element(arr.begin(), arr.end()); // 得到数组中的最大值
        int diff = target; // 计算目前的差距，初始值为最大的可能差距

        // 答案会在1-max之间
        for(int i = 1; i <= max; i++) {
            int loc = lower_bound(arr.begin(), arr.end(), i) - arr.begin(); // 找到大于或者等于i的第一个位置
            int cur = loc == 0 ? arr.size() * i :  prefix[loc - 1] + (arr.size() - loc) * i; // 如果替换后的数值大小
            //cout << "cur:" << cur << ", loc "<< loc << " i:" << i << " " << ", prefix:" << prefix[loc];
            if(abs(cur - target) < diff) {
                diff = abs(cur - target);
                ans = i;
            }
        }
        return ans; 
    }
};